### ISSUE TITLE
Serializing Error in serializer.py

### ISSUE THREAD

**Issue by sbidy** [adamghill/django-unicorn#132]
I get the following error if I pass this `{"id":"NMddZ6vg:requirement-view:1","data":{"is_editing":false,"requirement":{"created":"2021-01-26T10:59:25.399Z","last_updated":"2021-01-26T10:59:28.908Z","name":"Test","description":"Description for Test","pk":1}},"checksum":"2JbJYULj","actionQueue":[{"type":"callMethod","payload":{"name":"edit"},"partial":{}}]}`
 POST request to the component.

```
Internal Server Error: /unicorn/message/requirement-view
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 179, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/decorators.py", line 59, in timed
    result = wrapped(*args, **kwargs)
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/views.py", line 29, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/decorators/http.py", line 40, in inner
    return func(request, *args, **kwargs)
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/views.py", line 357, in message
    component_request.data = orjson.loads(component.get_frontend_context_variables())
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/decorators.py", line 59, in timed
    result = wrapped(*args, **kwargs)
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/components.py", line 409, in get_frontend_context_variables
    encoded_frontend_context_variables = serializer.dumps(
  File "/app/.local/lib/python3.8/site-packages/django_unicorn/serializer.py", line 127, in dumps
    serialized_data = orjson.dumps(data, default=_json_serializer)
TypeError: Type is not JSON serializable: Requirement
```

After some testing and debugging I recognized that the DateTime fields (`last_updated` and `created`) are the possible pain points.
I made some "debug print()" in the _dumps function
```python
    print(data["requirement"])
    serialized_data = orjson.dumps(data, default=_json_serializer)
    print(serialized_data)
    if fix_floats:
        return _dumps(serialized_data)

    return serialized_data
```

and got the following output:

1st print:
```
{'is_editing': False, 'requirement': {'created': '2021-01-26T10:59:25.399Z', 'last_updated': '2021-01-26T10:59:28.908Z', 'name': 'Test', 'description': 'Description for Test', 'pk': 1}}
```
2nd print:
```
b'{"is_editing":false,"requirement":{"created":"2021-01-26T10:59:25.399Z","last_updated":"2021-01-26T10:59:28.908Z","name":"Test","description":"Description for Test","pk":1}}'
```
3rd was the recursive call from `fix_floats` :
```
{'is_editing': True, 'requirement': <Requirement: Test>}
```
After the 3rd comes the crash - in that case the "requirement" was serialized as Requirement object from my Requirement model.

Can you maybe reproduce the serializing error with this POST data?

If I remove the DateTime fields it works.

In my model I use the `created = models.DateTimeField(auto_now_add=True, editable=False)`
I can also change the entry in the admin view.

**Comment by adamghill** [comment]
`fix_floats` is only _supposed_ convert `float` types to strings for JSON, but sounds like there might be a bug.

Just to make sure I understand `Requirement` is a Django model (i.e. it inherits from `django.db.models.Model`)?

I'll try to reproduce on my own, but if you could write a breaking unit test in this project, fork this project and change a component in the example to break in the same way, or copy/paste your `UnicornView` that is breaking into this issue that would be helpful. Thanks!

**Comment by sbidy** [comment]
My `UnicornView` is really basic:

```python
from django_unicorn.components import UnicornView

class RequirementViewView(UnicornView):
    requirement = None
    is_editing = False

    def edit(self):
        self.is_editing = True

    def cancel(self):
        self.is_editing = False

    def delete(self):
        self.requirement.delete()
        self.is_editing = False

    def save(self):
        self.requirement.save()
        self.is_editing = False
```

and the parent:
```python
from django_unicorn.components import UnicornView
from requirements.models import Requirement

class RequirementTableView(UnicornView):
    requirements = Requirement.objects.none()

    def mount(self):
        self.load_table()

    def load_table(self):
        self.requirements = Requirement.objects.all()

        for child in self.children:
            if hasattr(child, "is_editing"):
                child.is_editing = False
```
I took this from you example.
I'm looking forward to create a unit test for the serializer with an model containing the TimeDate field.

**Comment by sbidy** [comment]
If created a test for that issue .... but this will not popup with an issue. 😭
I'm not sure but maybe the `.decode("utf-8")` in the `dumps` made the change? I don't see this in the current published release.

In my above posted debug print all looks good -> `{'is_editing': True, 'requirement': <Requirement: Test>}` because `'requirement': <Requirement: Test>` is the object from my model `Requirement`.

**Comment by adamghill** [comment]
Hmm, `.decode("utf-8")` did change recently, but based on your stacktrace I don't think that's the problem. I used your code in PR #133 to create add a new test in b9a3f7e, but even running that test on the code tagged at 0.17.0 I don't see any issues.

However, I am able to replicate the issue in my example project, so I will be able to dig into this and figure out what is going on. Thanks for helping me finding the problem -- I'll get this fixed as soon as possible!

**Comment by adamghill** [comment]
This was related to having a model with a `DateTimeField` where the value was actually a string (not a `datetime`) and when Django tried to serialize the model it was raising an error when calling `isoformat` on the string.

I ended up looking at all fields of the model and if it's a `DateTimeField`, `DateField`, `TimeField`, or `DurationField` using the built-in Django utility functions to parse the string into the appropriate type in commit 5f7aaaed33901703e03ccdd54ff5a6f42a5f20d3.

Not sure if there is a better solution to this or other field types that might break this in the future, but hopefully this should solve this particular issue.

I'm going to close this for now, but feel free to open a new issue if you run into anything else!

**Comment by adamghill** [comment]
Fix released in 0.17.2. https://www.django-unicorn.com/docs/changelog/#v0-17-2

**Comment by sbidy** [comment]
Yes it works now with the 0.17.2 😉 ... Thank you for the fast response!
